program-runtime/src/executor_cache.rs

```diff
@@ -14,80 +14,69 @@ use {
-pub enum TxBankExecutorCacheDiff {
-    /// The TransactionExecutorCacheEntry did not change and is the same as the BankExecutorCacheEntry.
-    None,
-    /// The TransactionExecutorCacheEntry was inserted, no matching BankExecutorCacheEntry exists, so it needs to be inserted.
-    Inserted,
-    /// The TransactionExecutorCacheEntry was replaced, the matching BankExecutorCacheEntry needs to be updated.
-    Updated,
-}
-
-/// An entry of the TransactionExecutorCache
-#[derive(Debug)]
-pub struct TransactionExecutorCacheEntry {
-    executor: Arc<dyn Executor>,
-    difference: TxBankExecutorCacheDiff,
-}
-
 /// A subset of the BankExecutorCache containing only the executors relevant to one transaction
 ///
 /// The BankExecutorCache can not be updated directly as transaction batches are
 /// processed in parallel, which would cause a race condition.
 #[derive(Default, Debug)]
 pub struct TransactionExecutorCache {
-    pub executors: HashMap<Pubkey, TransactionExecutorCacheEntry>,
+    /// Executors which are visible to the transaction
+    pub executors: HashMap<Pubkey, Option<Arc<dyn Executor>>>,
+    /// Executors of programs which were deploymed during the transaction
+    pub deployments: HashMap<Pubkey, Arc<dyn Executor>>,
+    /// Executors which were loaded after the transaction was loaded (because they were missing in the cache)
+    pub loaded: HashMap<Pubkey, Arc<dyn Executor>>,
 }
 
 impl TransactionExecutorCache {
     pub fn new(executable_keys: impl Iterator<Item = (Pubkey, Arc<dyn Executor>)>) -> Self {
         Self {
             executors: executable_keys
-                .map(|(key, executor)| {
-                    let entry = TransactionExecutorCacheEntry {
-                        executor,
-                        difference: TxBankExecutorCacheDiff::None,
-                    };
-                    (key, entry)
-                })
+                .map(|(id, executor)| (id, Some(executor)))
                 .collect(),
+            deployments: HashMap::new(),
+            loaded: HashMap::new(),
         }
     }
 
-    pub fn get(&self, key: &Pubkey) -> Option<Arc<dyn Executor>> {
-        self.executors.get(key).map(|entry| entry.executor.clone())
+    pub fn get(&self, key: &Pubkey) -> Option<Option<Arc<dyn Executor>>> {
+        self.executors.get(key).cloned()
     }
 
-    pub fn set(&mut self, key: Pubkey, executor: Arc<dyn Executor>, replacement: bool) {
-        let difference = if replacement {
-            TxBankExecutorCacheDiff::Updated
+    pub fn set(
+        &mut self,
+        key: Pubkey,
+        executor: Arc<dyn Executor>,
+        upgrade: bool,
+        delay_visibility_of_program_deployment: bool,
+    ) {
+        if upgrade {
+            if !delay_visibility_of_program_deployment {
+                self.executors.insert(key, Some(executor.clone()));
+            } else {
+                // Do not overwrite an existing entry
+                self.executors.entry(key).or_insert(None);
```

Comment:
Why insert `None` if not found?

Reply:
That is essentially an explicit invalid entry similar to a tombstone. So that if a program was not deployed before the transaction begins, and it is deployed during the transaction, that the rest of the transaction still sees it as undeployed. In other words it is necessary to get case 1 (deployment invisibility) right.